### DISCUSSION TITLE
ALE is impeding undo with clang-format fixer and g:ale_fix_on_save = 1

### DISCUSSION THREAD

**Discussion by sdondley** [w0rp/ale#4564] [Q&A - Ask for help with problems]
I have this c code:

```
  #include <ctype.h>

  /* atof: convert string s to double */
  double atof(char s[]) {

<snip>

    return sign * val / power;
  }
```

I have this in my vim file:

```
  let g:ale_linters = {
      \ 'python': ['pylint'],
      \ 'vim': ['vint'],
      \ 'cpp': ['clang'],
      "\ 'c': ['clang'],
      \ 'markdown': ['languagetool']
  \}

  let g:ale_fixers = {
  \ 'c': ['clang-format'],
  \}
  let g:ale_fix_on_save = 1

  " To prevent buffer from resizing when loclist is updated
  let g:ale_set_loclist = 0

  " To prevent the cursor from moving when errors are found
  let g:ale_cursor_detail = 0
  set signcolumn=yes
```

If I remove the "r" in return at the bottom of the function, I get an error as expected. But when I try to undo to fix the problem, the line temporarily changes and then reverts back to what it was with the same error.

If I change the return line to simply `return 0` and delete the "r" and do undo, the "r" is reverted back and the error disappears. So there's something about `return sign * val / power;`that confuses ALE with fix on save turned on. Any fix aside from turning off fix on save?

**Comment by sdondley** [comment]
Until I can find a better solution, I'm doing this for a workaround:

`nnoremap <leader>ul :let g:ale_enabled = 0<CR> :undo<CR> :undo<CR> :let g:ale_enabled = 1<CR>`

**Comment by w0rp** [comment]
I'd report this as a bug. This sounds strange. Do you have Vim configured to automatically save files? I'm not sure what would cause this behaviour other than a fixer running.
